File: android/src/main/java/org/maplibre/reactnative/components/location/MLRNNativeUserLocationManager.java
package org.maplibre.reactnative.components.location;

import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactProp;
import org.maplibre.android.location.modes.RenderMode;

import javax.annotation.Nonnull;

public class MLRNNativeUserLocationManager extends ViewGroupManager<MLRNNativeUserLocation> {
    public static final String REACT_CLASS = "MLRNNativeUserLocation";

    @Nonnull
    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @ReactProp(name="androidRenderMode")
    public void setAndroidRenderMode(MLRNNativeUserLocation userLocation, String mode) {
       if ("compass".equalsIgnoreCase(mode)) {
           userLocation.setRenderMode(RenderMode.COMPASS);
        } else if ("gps".equalsIgnoreCase(mode)) {
           userLocation.setRenderMode(RenderMode.GPS);
        } else {
           userLocation.setRenderMode(RenderMode.NORMAL);
       }
    }

    @ReactProp(name="androidPreferredFramesPerSecond")
    public void setPreferredFramesPerSecond(MLRNNativeUserLocation userLocation, int preferredFramesPerSecond) {
       userLocation.setPreferredFramesPerSecond(preferredFramesPerSecond);
    }

    @Nonnull
    @Override
    protected MLRNNativeUserLocation createViewInstance(@Nonnull ThemedReactContext reactContext) {
        return new MLRNNativeUserLocation(reactContext);
    }
}

File: package/android/src/main/java/org/maplibre/reactnative/components/location/MLRNNativeUserLocationManager.kt
package org.maplibre.reactnative.components.location

import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewGroupManager
import com.facebook.react.uimanager.ViewManagerDelegate
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.viewmanagers.MLRNNativeUserLocationManagerDelegate
import com.facebook.react.viewmanagers.MLRNNativeUserLocationManagerInterface
import org.maplibre.android.location.modes.RenderMode

@ReactModule(name = MLRNNativeUserLocationManager.REACT_CLASS)
class MLRNNativeUserLocationManager :
    ViewGroupManager<MLRNNativeUserLocation>(),
    MLRNNativeUserLocationManagerInterface<MLRNNativeUserLocation> {
    private val delegate: MLRNNativeUserLocationManagerDelegate<MLRNNativeUserLocation, MLRNNativeUserLocationManager> =
        MLRNNativeUserLocationManagerDelegate(this)

    override fun getDelegate(): ViewManagerDelegate<MLRNNativeUserLocation> = delegate

    companion object {
        const val REACT_CLASS: String = "MLRNNativeUserLocation"
    }

    override fun getName(): String = REACT_CLASS

    override fun createViewInstance(context: ThemedReactContext): MLRNNativeUserLocation = MLRNNativeUserLocation(context)

    @ReactProp(name = "mode")
    override fun setMode(
        userLocation: MLRNNativeUserLocation,
        mode: String?,
    ) {
        if ("heading".equals(mode, ignoreCase = true)) {
            userLocation.setRenderMode(RenderMode.COMPASS)
        } else if ("course".equals(mode, ignoreCase = true)) {
            userLocation.setRenderMode(RenderMode.GPS)
        } else {
            userLocation.setRenderMode(RenderMode.NORMAL)
        }
    }

    @ReactProp(name = "androidPreferredFramesPerSecond")
    override fun setAndroidPreferredFramesPerSecond(
        userLocation: MLRNNativeUserLocation,
        preferredFramesPerSecond: Int,
    ) {
        userLocation.setPreferredFramesPerSecond(preferredFramesPerSecond)
    }
}
